#  Pure spin-orbit coupling

Mix the states with the same $L$ (and $m_L$) and the same elctron configuration $\mid 5s5p \rangle$

LHS: mixed spin-orbit states, RHS: pure spin-orbit states

$\mid ^3 P_0 \rangle = \mid ^3 P^0_0 \rangle$

$\mid ^3 P_1 \rangle = \alpha \mid ^3 P_1^0 \rangle + \beta \mid ^1 P_1^0 \rangle$

$\mid ^3 P_2 \rangle = \mid ^3 P_2^0 \rangle$

$\mid ^1 P_1 \rangle = -\beta \mid ^3 P_1^0 \rangle + \alpha \mid ^1P_1^0 \rangle$

The intermediate coupling coefficients are determined by

$\frac{\alpha^2}{\beta^2} = \frac{\tau^{^3P_1}}{\tau^{^1P_1}} (\frac{\nu^{^3P1}}{\nu^{^1P1}})^3$, $\alpha^2 + \beta^2 = 1$

In [1]:
# Ratio of (alpha/beta)^2, according to Stellmer
from math import sqrt
from __future__ import division
c = (434829 / 650504)**3 * (30.5 / 0.0074)

# The beta coefficient is negative, according to Boyd 2007
beta = - sqrt(1 / (1 + c))
alpha = sqrt(c / (1 + c))
print (alpha, beta)

(0.9995940880788805, -0.028489631056776547)


# Hyperfine coupling

The magnetic dipole ($A$) and electric quadrupole ($Q$) contributions to the hyperfine interaction are

$H_A = A \vec{I} \cdot \vec{J}$

$H_Q = Q \frac{\frac{3}{2} \vec{I} \cdot \vec{J} (2 \vec{I} \cdot \vec{J} + 1) - IJ(I+1) (J+1)}{2IJ(2I-1)(2J-1)}$

Basically they couple states with the same $F$ ($\vec{F} = \vec{I} + \vec{J}$) and $m_F$. For these excited states,


$\mid ^3 P_1 \rangle: A = -260$ MHz, $Q = -35 $MHz

$\mid ^3 P_2 \rangle: A = -212 $ MHz, $Q = 67 $MHz 

$\mid ^1 P_1 \rangle: A = -3.4 $ MHz, $Q = 39$ MHz 

In [2]:
# In units of MHz
A_3P1 = -260
Q_3P1 = -35
A_3P2 = -212
Q_3P2 = 67
A_1P1 = -3.4
Q_1P1 = 39

# Zeeman interaction with external magnetic field

$H_z = (g_S S_z + g_L L_z - g_I I_z) \mu_0 B$

where $g_S = 2, g_L = 1, g_I = \frac{\mu_I (1-\sigma_d)}{\mu_0 I}$

For $^{87} Sr$, $\mu_I = -1.09247 \mu_N, \sigma_d = 0.00345$ ($\sigma_d$ diamagnetic correction)

Then the full treatment including Zeeman interactions and HFI is

$H = H_Z + H_A + H_Q$

we assume these interactions are smaller than the fine structure such that we can use the spin-orbit mixed states as a basis.

# Clebsch-Gordan coefficients

## 1. Fine structure

We first express the pure spin-orbit states $\mid ^{2S + 1} P _ J ^0, m_J \rangle$ as a superposition of $\mid ^{2S + 1} P _ J ^0, m_S, m_L \rangle$ states using CG coefficients

$\mid ^{2S + 1} P _ J ^0, m_J \rangle = \sum_{m_S, m_L} C^{J, m_J}_{S, m_S, L, m_L} \mid ^{2S + 1} P _ J ^0, m_S, m_L \rangle$

Explicitly, we have

$\mid ^{3} P _ 0 ^0,  0\rangle = \mid ^{3} P _ 0 ^0, 0, 0 \rangle$


--------------------

$\mid ^{3} P _ 1 ^0, 1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, 1 \rangle)$

$\mid ^{3} P _ 1 ^0, 0 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, -1 \rangle - \mid ^{3} P _ 1 ^0, -1, 1 \rangle)$

$\mid ^{3} P _ 1 ^0, -1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, -1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, -1 \rangle)$

------------------

$\mid ^{3} P _ 2 ^0, 2 \rangle = \mid ^{3} P _ 2 ^0, 1, 1 \rangle $

$\mid ^{3} P _ 2 ^0, 1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 2 ^0, 1, 0 \rangle + \mid ^{3} P _ 2 ^0, 0, 1 \rangle)$

$\mid ^{3} P _ 2 ^0, 0 \rangle = \frac{\sqrt{2}}{\sqrt{3}}\mid ^{3} P _ 2 ^0, 0, 0 \rangle + \frac{1}{\sqrt{6}}\mid ^{3} P _ 2 ^0, 1, -1 \rangle + \frac{1}{\sqrt{6}}\mid ^{3} P _ 2 ^0, -1, 1 \rangle$

$\mid ^{3} P _ 2 ^0, -1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 2 ^0, -1, 0 \rangle + \mid ^{3} P _ 2 ^0, 0, -1 \rangle)$

$\mid ^{3} P _ 2 ^0, -2 \rangle = \mid ^{3} P _ 2 ^0, -1, -1 \rangle $

--------------------

$\mid ^{1} P _ 1 ^0, 1 \rangle = \mid ^{1} P _ 1 ^0, 0, 1 \rangle $

$\mid ^{1} P _ 1 ^0, 0 \rangle = \mid ^{1} P _ 1 ^0, 0, 1 \rangle $

$\mid ^{1} P _ 1 ^0, -1 \rangle = \mid ^{1} P _ 1 ^0, 0, -1 \rangle$

For the mixed spin-orbit states, we have

$\mid ^{3} P _ 0,  0\rangle = \mid ^{3} P _ 0 ^0, 0, 0 \rangle$


--------------------

$\mid ^{3} P _ 1, 1 \rangle = \frac{\alpha}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, 1 \rangle) + \beta \mid ^{1} P _ 1 ^0, 0, 1 \rangle$

$\mid ^{3} P _ 1, 0 \rangle = \frac{\alpha}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, -1 \rangle - \mid ^{3} P _ 1 ^0, -1, 1 \rangle) + \beta \mid ^{1} P _ 1 ^0, 0, 0 \rangle$

$\mid ^{3} P _ 1, -1 \rangle = \frac{\alpha}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, -1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, -1 \rangle) + \beta \mid ^{1} P _ 1 ^0, 0, -1 \rangle$

------------------

$\mid ^{3} P _ 2 , 2 \rangle = \mid ^{3} P _ 2 ^0, 1, 1 \rangle $

$\mid ^{3} P _ 2 , 1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 2 ^0, 1, 0 \rangle + \mid ^{3} P _ 2 ^0, 0, 1 \rangle)$

$\mid ^{3} P _ 2 , 0 \rangle = \frac{\sqrt{2}}{\sqrt{3}}\mid ^{3} P _ 2 ^0, 0, 0 \rangle + \frac{1}{\sqrt{6}}\mid ^{3} P _ 2 ^0, 1, -1 \rangle + \frac{1}{\sqrt{6}}\mid ^{3} P _ 2 ^0, -1, 1 \rangle$

$\mid ^{3} P _ 2 , -1 \rangle = \frac{1}{\sqrt{2}}(\mid ^{3} P _ 2 ^0, -1, 0 \rangle + \mid ^{3} P _ 2 ^0, 0, -1 \rangle)$

$\mid ^{3} P _ 2 , -2 \rangle = \mid ^{3} P _ 2 ^0, -1, -1 \rangle $

--------------------

$\mid ^{1} P _ 1, 1 \rangle = -\frac{\beta}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, 1 \rangle) + \alpha \mid ^{1} P _ 1 ^0, 0, 1 \rangle$

$\mid ^{1} P _ 1, 0 \rangle = -\frac{\beta}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, 1, -1 \rangle - \mid ^{3} P _ 1 ^0, -1, 1 \rangle) + \alpha \mid ^{1} P _ 1 ^0, 0, 0 \rangle$

$\mid ^{1} P _ 1, -1 \rangle = -\frac{\beta}{\sqrt{2}}(\mid ^{3} P _ 1 ^0, -1, 0 \rangle - \mid ^{3} P _ 1 ^0, 0, -1 \rangle) + \alpha \mid ^{1} P _ 1 ^0, 0, -1 \rangle$

In [3]:
from sympy.physics.quantum.cg import CG
from sympy import S
for s in [-1, 0, 1]:
    cg = CG(1, s, S(9)/2, S(3)/2 - s, S(11)/2, S(3)/2)
    print cg.doit()

sqrt(330)/55
2*sqrt(385)/55
sqrt(1155)/55


## 2. Hyperfine structure

Then we have

$\mid ^{2S + 1} P _ J , F, m_F \rangle = \sum_{m_J, m_I} C^{F, m_F}_{J, m_J, I, m_I} \mid ^{2S + 1} P _ J, m_J, m_I \rangle$

Explicitly, we have


$\mid ^{3} P _ 0, \frac{9}{2}, m_F\rangle = \mid ^{3} P _ 0, 0, m_F \rangle$


--------------------

$\mid ^{3} P _ 1, \frac{11}{2}, \frac{11}{2}\rangle = \mid ^{3} P _ 1, 1, \frac{9}{2} \rangle$

$\mid ^{3} P _ 1, \frac{11}{2}, \frac{9}{2}\rangle = \frac{\sqrt{2}}{\sqrt{11}}\mid ^{3} P _ 1, 0, \frac{9}{2} \rangle + \frac{3}{\sqrt{11}} \mid ^{3} P _ 1, 1, \frac{7}{2} \rangle $

$\mid ^{3} P _ 1, \frac{11}{2}, \frac{7}{2}\rangle = \frac{1}{\sqrt{55}}\mid ^{3} P _ 1, -1, \frac{9}{2} \rangle + \frac{3\sqrt{2}}{\sqrt{55}} \mid ^{3} P _ 1, 0, \frac{7}{2} \rangle+ \frac{6}{\sqrt{55}} \mid ^{3} P _ 1, 1, \frac{5}{2} \rangle  $

$\mid ^{3} P _ 1, \frac{11}{2}, \frac{5}{2}\rangle = \frac{\sqrt{3}}{\sqrt{55}}\mid ^{3} P _ 1, -1, \frac{7}{2} \rangle + \frac{2\sqrt{6}}{\sqrt{55}} \mid ^{3} P _ 1, 0, \frac{5}{2} \rangle+ \frac{2 \sqrt{7}}{\sqrt{55}} \mid ^{3} P _ 1, 1, \frac{3}{2} \rangle  $

$\mid ^{3} P _ 1, \frac{11}{2}, \frac{3}{2}\rangle = \frac{\sqrt{6}}{\sqrt{55}}\mid ^{3} P _ 1, -1, \frac{5}{2} \rangle + \frac{2\sqrt{7}}{\sqrt{55}} \mid ^{3} P _ 1, 0, \frac{3}{2} \rangle+ \frac{\sqrt{21}}{\sqrt{55}} \mid ^{3} P _ 1, 1, \frac{1}{2} \rangle  $

....

Too complicated

### $^3P_1$ states

In [18]:
# CG coefficients for m_J + m_I = 11/2
# m_J = -1, 0, 1
# m_F = -11/2
#       -9/2
#       ...

from sympy.physics.quantum.cg import CG
from sympy import S
from numpy import zeros

CG_3P1_11 = zeros((12,3))

# J, I, m_J, m_I, m_F
Eigen_3P1_11 = zeros((12,3),dtype='S3, d, d, d, d, d')

for m, i in enumerate(range(-11, 13, 2)):
    for n, j in enumerate(range(-1, 2, 1)):
        cg = CG(1, j, S(9)/2, S(i)/2 - j, S(11)/2, S(i)/2)
        CG_3P1_11[m][n] = cg.doit()
        Eigen_3P1_11[m][n][0] = '3P1'
        Eigen_3P1_11[m][n][1] = 9/2
        Eigen_3P1_11[m][n][2] = j
        Eigen_3P1_11[m][n][3] = i/2
        Eigen_3P1_11[m][n][4] = j + i/2
        
print '3P1 F = 11/2'  
print CG_3P1_11
print Eigen_3P1_11

3P1 F = 11/2
[[ 1.          0.          0.        ]
 [ 0.90453403  0.42640143  0.        ]
 [ 0.80903983  0.57207755  0.13483997]
 [ 0.71350607  0.66057826  0.23354968]
 [ 0.61791438  0.71350607  0.33028913]
 [ 0.52223297  0.73854895  0.42640143]
 [ 0.42640143  0.73854895  0.52223297]
 [ 0.33028913  0.71350607  0.61791438]
 [ 0.23354968  0.66057826  0.71350607]
 [ 0.13483997  0.57207755  0.80903983]
 [ 0.          0.42640143  0.90453403]
 [ 0.          0.          1.        ]]
[[('3P1', 4.5, -1.0, -5.5, -6.5, 0.0) ('3P1', 4.5, 0.0, -5.5, -5.5, 0.0)
  ('3P1', 4.5, 1.0, -5.5, -4.5, 0.0)]
 [('3P1', 4.5, -1.0, -4.5, -5.5, 0.0) ('3P1', 4.5, 0.0, -4.5, -4.5, 0.0)
  ('3P1', 4.5, 1.0, -4.5, -3.5, 0.0)]
 [('3P1', 4.5, -1.0, -3.5, -4.5, 0.0) ('3P1', 4.5, 0.0, -3.5, -3.5, 0.0)
  ('3P1', 4.5, 1.0, -3.5, -2.5, 0.0)]
 [('3P1', 4.5, -1.0, -2.5, -3.5, 0.0) ('3P1', 4.5, 0.0, -2.5, -2.5, 0.0)
  ('3P1', 4.5, 1.0, -2.5, -1.5, 0.0)]
 [('3P1', 4.5, -1.0, -1.5, -2.5, 0.0) ('3P1', 4.5, 0.0, -1.5, -1.5, 0.0)
 

In [14]:
# CG coefficients for m_J + m_I = 9/2
# m_J = -1, 0, 1
# m_F = -9/2
#       -7/2
#       ...

CG_3P1_9 = zeros((10,3))

# J, I, m_J, m_I, m_F
Eigen_3P1_9 = zeros((10,3),dtype='d, d, d, d, d')

for m, i in enumerate(range(-9, 11, 2)):
    for n, j in enumerate(range(-1, 2, 1)):
        cg = CG(1, j, S(9)/2, S(i)/2 - j, S(9)/2, S(i)/2)
        CG_3P1_9[m][n] = cg.doit()
        Eigen_3P1_9[m][n][0] = 1
        Eigen_3P1_9[m][n][1] = 9/2
        Eigen_3P1_9[m][n][2] = j
        Eigen_3P1_9[m][n][3] = i/2
        Eigen_3P1_9[m][n][4] = j + i/2

print '3P1 F = 9/2'
print CG_3P1_9
print Eigen_3P1_9

3P1 F = 9/2
[[-0.42640143  0.90453403  0.        ]
 [-0.56853524  0.70352647  0.42640143]
 [-0.65133895  0.50251891  0.56853524]
 [-0.69631062  0.30151134  0.65133895]
 [-0.71066905  0.10050378  0.69631062]
 [-0.69631062 -0.10050378  0.71066905]
 [-0.65133895 -0.30151134  0.69631062]
 [-0.56853524 -0.50251891  0.65133895]
 [-0.42640143 -0.70352647  0.56853524]
 [ 0.         -0.90453403  0.42640143]]
[[(1.0, 4.5, -1.0, -4.5, -5.5) (1.0, 4.5, 0.0, -4.5, -4.5)
  (1.0, 4.5, 1.0, -4.5, -3.5)]
 [(1.0, 4.5, -1.0, -3.5, -4.5) (1.0, 4.5, 0.0, -3.5, -3.5)
  (1.0, 4.5, 1.0, -3.5, -2.5)]
 [(1.0, 4.5, -1.0, -2.5, -3.5) (1.0, 4.5, 0.0, -2.5, -2.5)
  (1.0, 4.5, 1.0, -2.5, -1.5)]
 [(1.0, 4.5, -1.0, -1.5, -2.5) (1.0, 4.5, 0.0, -1.5, -1.5)
  (1.0, 4.5, 1.0, -1.5, -0.5)]
 [(1.0, 4.5, -1.0, -0.5, -1.5) (1.0, 4.5, 0.0, -0.5, -0.5)
  (1.0, 4.5, 1.0, -0.5, 0.5)]
 [(1.0, 4.5, -1.0, 0.5, -0.5) (1.0, 4.5, 0.0, 0.5, 0.5)
  (1.0, 4.5, 1.0, 0.5, 1.5)]
 [(1.0, 4.5, -1.0, 1.5, 0.5) (1.0, 4.5, 0.0, 1.5, 1.5)
  (1.0, 

In [6]:
# CG coefficients for m_J + m_I = 7/2
# m_J = -1, 0, 1
# m_F = -7/2
#       -5/2
#       ...

CG_3P1_7 = zeros((8,3))

# J, I, m_J, m_I, m_F
Eigen_3P1_7 = zeros((8,3),dtype='d, d, d, d, d')

for m, i in enumerate(range(-7, 9, 2)):
    for n, j in enumerate(range(-1, 2, 1)):
        cg = CG(1, j, S(9)/2, S(i)/2 - j, S(7)/2, S(i)/2)
        CG_3P1_7[m][n] = cg.doit()
        Eigen_3P1_7[m][n][0] = 1
        Eigen_3P1_7[m][n][1] = 7/2
        Eigen_3P1_7[m][n][2] = j
        Eigen_3P1_7[m][n][3] = i/2
        Eigen_3P1_7[m][n][4] = j + i/2
        
print '3P1 F = 7/2'
print CG_3P1_7
print Eigen_3P1_7

3P1 F = 7/2
[[ 0.1490712  -0.42163702  0.89442719]
 [ 0.25819889 -0.55777335  0.78881064]
 [ 0.36514837 -0.63245553  0.68313005]
 [ 0.47140452 -0.66666667  0.57735027]
 [ 0.57735027 -0.66666667  0.47140452]
 [ 0.68313005 -0.63245553  0.36514837]
 [ 0.78881064 -0.55777335  0.25819889]
 [ 0.89442719 -0.42163702  0.1490712 ]]
[[(1.0, 3.5, -1.0, -3.5, -4.5) (1.0, 3.5, 0.0, -3.5, -3.5)
  (1.0, 3.5, 1.0, -3.5, -2.5)]
 [(1.0, 3.5, -1.0, -2.5, -3.5) (1.0, 3.5, 0.0, -2.5, -2.5)
  (1.0, 3.5, 1.0, -2.5, -1.5)]
 [(1.0, 3.5, -1.0, -1.5, -2.5) (1.0, 3.5, 0.0, -1.5, -1.5)
  (1.0, 3.5, 1.0, -1.5, -0.5)]
 [(1.0, 3.5, -1.0, -0.5, -1.5) (1.0, 3.5, 0.0, -0.5, -0.5)
  (1.0, 3.5, 1.0, -0.5, 0.5)]
 [(1.0, 3.5, -1.0, 0.5, -0.5) (1.0, 3.5, 0.0, 0.5, 0.5)
  (1.0, 3.5, 1.0, 0.5, 1.5)]
 [(1.0, 3.5, -1.0, 1.5, 0.5) (1.0, 3.5, 0.0, 1.5, 1.5)
  (1.0, 3.5, 1.0, 1.5, 2.5)]
 [(1.0, 3.5, -1.0, 2.5, 1.5) (1.0, 3.5, 0.0, 2.5, 2.5)
  (1.0, 3.5, 1.0, 2.5, 3.5)]
 [(1.0, 3.5, -1.0, 3.5, 2.5) (1.0, 3.5, 0.0, 3.5, 3.5)
  (1.0, 

### $^3P_2$ states

In [7]:
# CG coefficients for m_J + m_I = 13/2
# m_J = -2, -1, 0, 1, 2
# m_F = -13/2
#       -11/2
#       ...

CG_3P2_13 = zeros((14,5))

# J, I, m_J, m_I, m_F
Eigen_3P2_13 = zeros((14,5),dtype='d, d, d, d, d')

for m, i in enumerate(range(-13, 15, 2)):
    for n, j in enumerate(range(-2, 3, 1)):
        cg = CG(2, j, S(9)/2, S(i)/2 - j, S(13)/2, S(i)/2)
        CG_3P2_13[m][n] = cg.doit()
        Eigen_3P2_13[m][n][0] = 2
        Eigen_3P2_13[m][n][1] = 13/2
        Eigen_3P2_13[m][n][2] = j
        Eigen_3P2_13[m][n][3] = i/2
        Eigen_3P2_13[m][n][4] = j + i/2
        
print '3P2 F = 13/2'  
print CG_3P2_13
print Eigen_3P2_13

3P2 F = 13/2
[[ 1.          0.          0.          0.          0.        ]
 [ 0.83205029  0.5547002   0.          0.          0.        ]
 [ 0.67936622  0.67936622  0.2773501   0.          0.        ]
 [ 0.54194676  0.70957488  0.43452409  0.11826248  0.        ]
 [ 0.41979016  0.68551446  0.54963433  0.22438728  0.03739788]
 [ 0.31289311  0.62578622  0.62578622  0.3344968   0.0836242 ]
 [ 0.22124884  0.54194676  0.66374652  0.44249768  0.14484136]
 [ 0.14484136  0.44249768  0.66374652  0.54194676  0.22124884]
 [ 0.0836242   0.3344968   0.62578622  0.62578622  0.31289311]
 [ 0.03739788  0.22438728  0.54963433  0.68551446  0.41979016]
 [ 0.          0.11826248  0.43452409  0.70957488  0.54194676]
 [ 0.          0.          0.2773501   0.67936622  0.67936622]
 [ 0.          0.          0.          0.5547002   0.83205029]
 [ 0.          0.          0.          0.          1.        ]]
[[(2.0, 6.5, -2.0, -6.5, -8.5) (2.0, 6.5, -1.0, -6.5, -7.5)
  (2.0, 6.5, 0.0, -6.5, -6.5) (2.0, 6.5, 1.0

In [8]:
# CG coefficients for m_J + m_I = 11/2
# m_J = -2, -1, 0, 1, 2
# m_F = -11/2
#       -9/2
#       ...

CG_3P2_11 = zeros((12,5))

# J, I, m_J, m_I, m_F
Eigen_3P2_11 = zeros((12,5),dtype='d, d, d, d, d')

for m, i in enumerate(range(-11, 13, 2)):
    for n, j in enumerate(range(-2, 3, 1)):
        cg = CG(2, j, S(9)/2, S(i)/2 - j, S(11)/2, S(i)/2)
        CG_3P2_11[m][n] = cg.doit()
        Eigen_3P2_11[m][n][0] = 2
        Eigen_3P2_11[m][n][1] = 11/2
        Eigen_3P2_11[m][n][2] = j
        Eigen_3P2_11[m][n][3] = i/2
        Eigen_3P2_11[m][n][4] = j + i/2
        
print '3P2 F = 11/2'  
print CG_3P2_11
print Eigen_3P2_11

3P2 F = 11/2
[[-0.5547002   0.83205029  0.          0.          0.        ]
 [-0.66899361  0.41812101  0.61450987  0.          0.        ]
 [-0.68551446  0.07479576  0.64124006  0.33658092  0.        ]
 [-0.64630923 -0.19789098  0.52888589  0.49660854  0.12955006]
 [-0.57126205 -0.39988343  0.34275723  0.58016958  0.24428193]
 [-0.47304992 -0.531085    0.11826248  0.5913124   0.36129784]
 [-0.36129784 -0.5913124  -0.11826248  0.531085    0.47304992]
 [-0.24428193 -0.58016958 -0.34275723  0.39988343  0.57126205]
 [-0.12955006 -0.49660854 -0.52888589  0.19789098  0.64630923]
 [ 0.         -0.33658092 -0.64124006 -0.07479576  0.68551446]
 [ 0.          0.         -0.61450987 -0.41812101  0.66899361]
 [ 0.          0.          0.         -0.83205029  0.5547002 ]]
[[(2.0, 5.5, -2.0, -5.5, -7.5) (2.0, 5.5, -1.0, -5.5, -6.5)
  (2.0, 5.5, 0.0, -5.5, -5.5) (2.0, 5.5, 1.0, -5.5, -4.5)
  (2.0, 5.5, 2.0, -5.5, -3.5)]
 [(2.0, 5.5, -2.0, -4.5, -6.5) (2.0, 5.5, -1.0, -4.5, -5.5)
  (2.0, 5.5, 0.0, -4.

In [9]:
# CG coefficients for m_J + m_I = 9/2
# m_J = -2, -1, 0, 1, 2
# m_F = -9/2
#       -7/2
#       ...

CG_3P2_9 = zeros((10,5))

# J, I, m_J, m_I, m_F
Eigen_3P2_9 = zeros((10,5),dtype='d, d, d, d, d')

for m, i in enumerate(range(-9, 11, 2)):
    for n, j in enumerate(range(-2, 3, 1)):
        cg = CG(2, j, S(9)/2, S(i)/2 - j, S(9)/2, S(i)/2)
        CG_3P2_9[m][n] = cg.doit()
        Eigen_3P2_9[m][n][0] = 2
        Eigen_3P2_9[m][n][1] = 9/2
        Eigen_3P2_9[m][n][2] = j
        Eigen_3P2_9[m][n][3] = i/2
        Eigen_3P2_9[m][n][4] = j + i/2
        
print '3P2 F = 9/2'  
print CG_3P2_9
print Eigen_3P2_9

3P2 F = 9/2
[[ 0.30151134 -0.60302269  0.73854895  0.          0.        ]
 [ 0.46056619 -0.60302269  0.24618298  0.60302269  0.        ]
 [ 0.56407607 -0.46056619 -0.12309149  0.60302269  0.30151134]
 [ 0.61545745 -0.24618298 -0.36927447  0.46056619  0.46056619]
 [ 0.61545745  0.         -0.49236596  0.24618298  0.56407607]
 [ 0.56407607  0.24618298 -0.49236596  0.          0.61545745]
 [ 0.46056619  0.46056619 -0.36927447 -0.24618298  0.61545745]
 [ 0.30151134  0.60302269 -0.12309149 -0.46056619  0.56407607]
 [ 0.          0.60302269  0.24618298 -0.60302269  0.46056619]
 [ 0.          0.          0.73854895 -0.60302269  0.30151134]]
[[(2.0, 4.5, -2.0, -4.5, -6.5) (2.0, 4.5, -1.0, -4.5, -5.5)
  (2.0, 4.5, 0.0, -4.5, -4.5) (2.0, 4.5, 1.0, -4.5, -3.5)
  (2.0, 4.5, 2.0, -4.5, -2.5)]
 [(2.0, 4.5, -2.0, -3.5, -5.5) (2.0, 4.5, -1.0, -3.5, -4.5)
  (2.0, 4.5, 0.0, -3.5, -3.5) (2.0, 4.5, 1.0, -3.5, -2.5)
  (2.0, 4.5, 2.0, -3.5, -1.5)]
 [(2.0, 4.5, -2.0, -2.5, -4.5) (2.0, 4.5, -1.0, -2.5, -3.5)

In [10]:
# CG coefficients for m_J + m_I = 7/2
# m_J = -2, -1, 0, 1, 2
# m_F = -7/2
#       -5/2
#       ...


CG_3P2_7 = zeros((8,5))

# J, I, m_J, m_I, m_F
Eigen_3P2_7 = zeros((8,5),dtype='d, d, d, d, d')

for m, i in enumerate(range(-7, 9, 2)):
    for n, j in enumerate(range(-2, 3, 1)):
        cg = CG(2, j, S(9)/2, S(i)/2 - j, S(7)/2, S(i)/2)
        CG_3P2_7[m][n] = cg.doit()
        Eigen_3P2_7[m][n][0] = 2
        Eigen_3P2_7[m][n][1] = 7/2
        Eigen_3P2_7[m][n][2] = j
        Eigen_3P2_7[m][n][3] = i/2
        Eigen_3P2_7[m][n][4] = j + i/2
        
print '3P2 F = 7/2'  
print CG_3P2_7
print Eigen_3P2_7

3P2 F = 7/2
[[-0.15569979  0.35675303 -0.58257527  0.71350607  0.        ]
 [-0.28829999  0.5002164  -0.55048188  0.26967994  0.53935989]
 [-0.41612519  0.54096275 -0.37451267 -0.07784989  0.62279916]
 [-0.52636136  0.48349378 -0.13159034 -0.32897585  0.60302269]
 [-0.60302269  0.32897585  0.13159034 -0.48349378  0.52636136]
 [-0.62279916  0.07784989  0.37451267 -0.54096275  0.41612519]
 [-0.53935989 -0.26967994  0.55048188 -0.5002164   0.28829999]
 [ 0.         -0.71350607  0.58257527 -0.35675303  0.15569979]]
[[(2.0, 3.5, -2.0, -3.5, -5.5) (2.0, 3.5, -1.0, -3.5, -4.5)
  (2.0, 3.5, 0.0, -3.5, -3.5) (2.0, 3.5, 1.0, -3.5, -2.5)
  (2.0, 3.5, 2.0, -3.5, -1.5)]
 [(2.0, 3.5, -2.0, -2.5, -4.5) (2.0, 3.5, -1.0, -2.5, -3.5)
  (2.0, 3.5, 0.0, -2.5, -2.5) (2.0, 3.5, 1.0, -2.5, -1.5)
  (2.0, 3.5, 2.0, -2.5, -0.5)]
 [(2.0, 3.5, -2.0, -1.5, -3.5) (2.0, 3.5, -1.0, -1.5, -2.5)
  (2.0, 3.5, 0.0, -1.5, -1.5) (2.0, 3.5, 1.0, -1.5, -0.5)
  (2.0, 3.5, 2.0, -1.5, 0.5)]
 [(2.0, 3.5, -2.0, -0.5, -2.5) (2.0, 

In [11]:
# CG coefficients for m_J + m_I = 5/2
# m_J = -2, -1, 0, 1, 2
# m_F = -5/2
#       -3/2
#       ...


CG_3P2_5 = zeros((6,5))

# J, I, m_J, m_I, m_F
Eigen_3P2_5 = zeros((6,5),dtype='d, d, d, d, d')

for m, i in enumerate(range(-5, 7, 2)):
    for n, j in enumerate(range(-2, 3, 1)):
        cg = CG(2, j, S(9)/2, S(i)/2 - j, S(5)/2, S(i)/2)
        CG_3P2_5[m][n] = cg.doit()
        Eigen_3P2_5[m][n][0] = 2
        Eigen_3P2_5[m][n][1] = 5/2
        Eigen_3P2_5[m][n][2] = j
        Eigen_3P2_5[m][n][3] = i/2
        Eigen_3P2_5[m][n][4] = j + i/2
        
print '3P2 F = 5/2'  
print CG_3P2_5
print Eigen_3P2_5

3P2 F = 5/2
[[ 0.06900656 -0.16903085  0.31622777 -0.51639778  0.77459667]
 [ 0.15430335 -0.3086067   0.46291005 -0.57735027  0.57735027]
 [ 0.26726124 -0.43643578  0.53452248 -0.53452248  0.40824829]
 [ 0.40824829 -0.53452248  0.53452248 -0.43643578  0.26726124]
 [ 0.57735027 -0.57735027  0.46291005 -0.3086067   0.15430335]
 [ 0.77459667 -0.51639778  0.31622777 -0.16903085  0.06900656]]
[[(2.0, 2.5, -2.0, -2.5, -4.5) (2.0, 2.5, -1.0, -2.5, -3.5)
  (2.0, 2.5, 0.0, -2.5, -2.5) (2.0, 2.5, 1.0, -2.5, -1.5)
  (2.0, 2.5, 2.0, -2.5, -0.5)]
 [(2.0, 2.5, -2.0, -1.5, -3.5) (2.0, 2.5, -1.0, -1.5, -2.5)
  (2.0, 2.5, 0.0, -1.5, -1.5) (2.0, 2.5, 1.0, -1.5, -0.5)
  (2.0, 2.5, 2.0, -1.5, 0.5)]
 [(2.0, 2.5, -2.0, -0.5, -2.5) (2.0, 2.5, -1.0, -0.5, -1.5)
  (2.0, 2.5, 0.0, -0.5, -0.5) (2.0, 2.5, 1.0, -0.5, 0.5)
  (2.0, 2.5, 2.0, -0.5, 1.5)]
 [(2.0, 2.5, -2.0, 0.5, -1.5) (2.0, 2.5, -1.0, 0.5, -0.5)
  (2.0, 2.5, 0.0, 0.5, 0.5) (2.0, 2.5, 1.0, 0.5, 1.5)
  (2.0, 2.5, 2.0, 0.5, 2.5)]
 [(2.0, 2.5, -2.0, 1.5, 

### $^1P_1$ states

Are exactly the same as the $^3 P _1$ states

In [12]:
CG_1P1_11 = CG_3P1_11
CG_1P1_9 = CG_3P1_9
CG_1P1_7 = CG_3P1_7

# Matrix elements of HFI

$ H_{HF} = A \vec{I} \cdot \vec{J} + Q \frac{\frac{3}{2} \vec{I} \cdot \vec{J} (2 \vec{I} \cdot \vec{J} + 1) - IJ(I+1) (J+1)}{2IJ(2I-1)(2J-1)}$

Using $\vec{I} \cdot \vec{J} = I_z J_z + \frac{1}{2} (I_+ J_- + I_- J_+)$

$J_+ \mid {J, m_J} \rangle = \hbar \sqrt{(J - m_J)(J + m_J + 1)} \mid {J, m_J + 1} \rangle$

$J_- \mid {J, m_J} \rangle = \hbar \sqrt{(J + m_J)(J - m_J + 1)} \mid {J, m_J - 1} \rangle$

and similar for $I_+, I_-$

$I_- J_+ \mid {J, I, m_J, m_I} \rangle = \sqrt{(I + m_I)(I - m_I + 1)}\sqrt{(J - m_J)(J + m_J + 1)} \mid {J, I, m_J + 1, m_I - 1} \rangle$

$I_+ J_- \mid {J, I, m_J, m_I} \rangle =  \sqrt{(I - m_I)(I + m_I + 1)}\sqrt{(J + m_J)(J - m_J + 1)} \mid {J, I, m_J - 1, m_I + 1} \rangle$

Then 

$\vec{I} \cdot \vec{J} \mid J, \frac{9}{2}, m_J, m_I \rangle = m_I m_J \mid {J, \frac{9}{2}, m_J, m_I} \rangle + \frac{1}{2}  \sqrt{(\frac{9}{2} - m_I)(\frac{9}{2} + m_I + 1)}\sqrt{(J + m_J)(J - m_J + 1)} \mid {J, \frac{9}{2}, m_J - 1, m_I + 1} \rangle + \frac{1}{2} \sqrt{(\frac{9}{2} + m_I)(\frac{9}{2} - m_I + 1)}\sqrt{(J - m_J)(J + m_J + 1)} \mid {J, \frac{9}{2}, m_J + 1, m_I - 1} \rangle$

$H_{HF}$ only couples states with the same $J, I, m_J + m_I = m_F$

$\langle J_1, m_{J_1}, m_{I_1}  \mid H_{HF} \mid J, m_{J}, m_{I} \rangle = (A + \frac{3}{2} Q \frac{1}{2IJ(2I-1)(2J-1)})(\frac{1}{2}  \sqrt{(\frac{9}{2} - m_I)(\frac{9}{2} + m_I + 1)}\sqrt{(J + m_J)(J - m_J + 1)} \delta_{J_1, J} \delta{m_{J_1}, m_{J} - 1} \delta{m_{I_1}, m_{I} + 1} + \frac{1}{2} \sqrt{(\frac{9}{2} + m_I)(\frac{9}{2} - m_I + 1)}\sqrt{(J - m_J)(J + m_J + 1)} \delta_{J_1, J} \delta{m_{J_1}, m_{J} + 1} \delta{m_{I_1}, m_{I} -1} + (A m_I m_J+  Q \frac{ \frac{3}{2}m_I m_J - IJ(I+1)(J+1)}{2IJ(2I-1)(2J-1)})\delta_{J_1, J} \delta{m_{J_1}, m_{J}} \delta{m_{I_1}, m_{I}} + 3 Q \frac{1}{2IJ(2I-1)(2J-1)}\vec{I} \cdot \vec{J} \vec{I} \cdot \vec{J}$


...

Too complicated

## 1. $m_F = -\frac{13}{2}$

The only possibility is $\mid ^{3} P _ 2, \frac{13}{2}, -\frac{13}{2}\rangle = \mid ^{3} P_ 2, -2, -\frac{9}{2}\rangle $

## 2. $m_F = -\frac{11}{2}$

The possibilities are

$\mid ^{3} P _ 2, \frac{13}{2}, -\frac{11}{2}\rangle = 0.83205029 \mid ^{3} P_ 2, -2, -\frac{7}{2}\rangle + 0.5547002  \mid ^{3} P_ 2, -1, -\frac{9}{2}\rangle $

$\mid ^{3} P _ 2, \frac{11}{2}, -\frac{11}{2}\rangle = -0.5547002 \mid ^{3} P_ 2, -2, -\frac{7}{2}\rangle + 0.83205029 \mid ^{3} P_ 2, -1, -\frac{9}{2}\rangle $

$\mid ^{3} P _ 1, \frac{11}{2}, -\frac{11}{2}\rangle = \mid ^{3} P_ 1, -1, -\frac{9}{2}\rangle $

$\mid ^{1} P _ 1, \frac{11}{2}, -\frac{11}{2}\rangle = \mid ^{3} P_ 1, -1, -\frac{9}{2}\rangle $